In [1]:
import pandas as pd
import matplotlib.pyplot as plt

In [2]:
%store -r ted

In [3]:
ted

comments                                        description  duration  \
0         4553  Sir Ken Robinson makes an entertaining and pro...      1164   
1          265  With the same humor and humanity he exuded in ...       977   
2          124  New York Times columnist David Pogue takes aim...      1286   
3          200  In an emotionally charged talk, MacArthur-winn...      1116   
4          593  You've never seen data presented like this. Wi...      1190   
...        ...                                                ...       ...   
2545        17  Between 2008 and 2016, the United States depor...       476   
2546         6  How can you study Mars without a spaceship? He...       290   
2547        10  Science fiction visions of the future show us ...       651   
2548        32  In an unmissable talk about race and politics ...      1100   
2549         8  With more than half of the world population li...       519   

             event   film_date  languages         main_speaker  \
0          TED2006  1140825600         60         Ken Robinson   
1          TED2006  1140825600         43              Al Gore   
2          TED2006  1140739200         26          David Pogue   
3          TED2006  1140912000         35        Majora Carter   
4          TED2006  1140566400         48         Hans Rosling   
...            ...         ...        ...                  ...   
2545       TED2017  1496707200          4     Duarte Geraldino   
2546       TED2017  1492992000          3  Armando Azua-Bustos   
2547       TED2017  1492992000          1       Radhika Nagpal   
2548  TEDxMileHigh  1499472000          1     Theo E.J. Wilson   
2549       TED2017  1492992000          1    Karoliina Korppoo   

                                                   name  num_speaker  \
0             Ken Robinson: Do schools kill creativity?            1   
1                  Al Gore: Averting the climate crisis            1   
2                         David Pogue: Simplicity sells            1   
3                    Majora Carter: Greening the ghetto            1   
4         Hans Rosling: The best stats you've ever seen            1   
...                                                 ...          ...   
2545  Duarte Geraldino: What we're missing in the de...            1   
2546  Armando Azua-Bustos: The most Martian place on...            1   
2547  Radhika Nagpal: What intelligent machines can ...            1   
2548  Theo E.J. Wilson: A black man goes undercover ...            1   
2549  Karoliina Korppoo: How a video game might help...            1   

      published_date                                            ratings  \
0         1151367060  [{'id': 7, 'name': 'Funny', 'count': 19645}, {...   
1         1151367060  [{'id': 7, 'name': 'Funny', 'count': 544}, {'i...   
2         1151367060  [{'id': 7, 'name': 'Funny', 'count': 964}, {'i...   
3         1151367060  [{'id': 3, 'name': 'Courageous', 'count': 760}...   
4         1151440680  [{'id': 9, 'name': 'Ingenious', 'count': 3202}...   
...              ...                                                ...   
2545      1505851216  [{'id': 3, 'name': 'Courageous', 'count': 24},...   
2546      1505919737  [{'id': 22, 'name': 'Fascinating', 'count': 32...   
2547      1506006095  [{'id': 1, 'name': 'Beautiful', 'count': 14}, ...   
2548      1506024042  [{'id': 11, 'name': 'Longwinded', 'count': 3},...   
2549      1506092422  [{'id': 21, 'name': 'Unconvincing', 'count': 2...   

                                          related_talks  \
0     [{'id': 865, 'hero': 'https://pe.tedcdn.com/im...   
1     [{'id': 243, 'hero': 'https://pe.tedcdn.com/im...   
2     [{'id': 1725, 'hero': 'https://pe.tedcdn.com/i...   
3     [{'id': 1041, 'hero': 'https://pe.tedcdn.com/i...   
4     [{'id': 2056, 'hero': 'https://pe.tedcdn.com/i...   
...                                                 ...   
2545  [{'id': 2596, 'hero': 'https://pe.tedcdn.com/i...   
2546  [{'id': 2491, 'hero'

# 5. What were the best events in TED history to attend?

It is our job to figure out how to define `"best"` and use that definition as our metric.

***First idea:*** check the total speakers at each event

***Follow up:*** Maybe the event had a lot of speakers but was not very popular. Look at the ratio of comments to views because the best way to evaluate such a quality is in hindsight so talks that have been viewed more and or commented more might be considered most interesting? ***Be careful*** controversial might not be considered the best and controversial also attracts a lot of comments....let's leave the comments out of this one and lot to pure view count....though we need to consider possibility that some views may not be intentional...however that might be more common with music streaming where people put it on in the background or play farming. Ted talks might be more deliberate viewings, not background nose. 

***Also:*** There is a column that contains ratings but the dtype looks like a list of dictionaries?

In [5]:
ted.ratings.value_counts().head()

[{'id': 3, 'name': 'Courageous', 'count': 255}, {'id': 8, 'name': 'Informative', 'count': 369}, {'id': 24, 'name': 'Persuasive', 'count': 341}, {'id': 10, 'name': 'Inspiring', 'count': 363}, {'id': 25, 'name': 'OK', 'count': 120}, {'id': 21, 'name': 'Unconvincing', 'count': 106}, {'id': 9, 'name': 'Ingenious', 'count': 42}, {'id': 23, 'name': 'Jaw-dropping', 'count': 35}, {'id': 11, 'name': 'Longwinded', 'count': 8}, {'id': 26, 'name': 'Obnoxious', 'count': 32}, {'id': 22, 'name': 'Fascinating', 'count': 97}, {'id': 1, 'name': 'Beautiful', 'count': 31}, {'id': 2, 'name': 'Confusing', 'count': 20}, {'id': 7, 'name': 'Funny', 'count': 3}]              1
[{'id': 1, 'name': 'Beautiful', 'count': 218}, {'id': 9, 'name': 'Ingenious', 'count': 730}, {'id': 7, 'name': 'Funny', 'count': 1021}, {'id': 3, 'name': 'Courageous', 'count': 284}, {'id': 21, 'name': 'Unconvincing', 'count': 36}, {'id': 11, 'name': 'Longwinded', 'count': 27}, {'id': 8, 'name': 'Informative', 'count': 1544}, {'id': 10, '

`Ratings` looks promising! We have inside each dictionary and id that is a numeric value pointing to an Adjective and then then number of times that adjective was used or checked, however tey collected the information. It seems like it is very organized so I imagine that the audience is given a list of adjectives that they must look through and check off all that apply. Providing a list would help reduce the variety of synonyms that would be more difficult to wrangle.
My concern is that there could be a lot of missing and irregular information here.

In [40]:
# First thing First just a simple count of the number od speaker at a Ted talk
ted.event.value_counts().sort_index(ascending=True).sample(10)

DLD 2007                 1
TEDxYouth@Manchester     2
TEDSummit               34
TEDxOslo                 2
TEDxPenn                 1
TEDxToronto 2011         2
TEDMED 2016             10
TEDxSF                   1
TED2001                  3
TEDxAmoskeagMillyard     1
Name: event, dtype: int64

In [39]:
# let's try to group these by event and sum the views and divide number of views by number of talks
ted_event_views = ted.groupby('event').agg({'event':'count', 'views':'sum'})

In [83]:
ted['views_per_speaker'] = ted_event_views.views/ted_event_views.event

In [ ]:
# We made a mistake by defining the new column in the original df.
# Here we are going to drop it.
ted.drop(columns='views_per_speaker')


In [91]:
ted.head()

comments                                        description  duration  \
0      4553  Sir Ken Robinson makes an entertaining and pro...      1164   
1       265  With the same humor and humanity he exuded in ...       977   
2       124  New York Times columnist David Pogue takes aim...      1286   
3       200  In an emotionally charged talk, MacArthur-winn...      1116   
4       593  You've never seen data presented like this. Wi...      1190   

     event   film_date  languages   main_speaker  \
0  TED2006  1140825600         60   Ken Robinson   
1  TED2006  1140825600         43        Al Gore   
2  TED2006  1140739200         26    David Pogue   
3  TED2006  1140912000         35  Majora Carter   
4  TED2006  1140566400         48   Hans Rosling   

                                            name  num_speaker  published_date  \
0      Ken Robinson: Do schools kill creativity?            1      1151367060   
1           Al Gore: Averting the climate crisis            1      1151367060   
2                  David Pogue: Simplicity sells            1      1151367060   
3             Majora Carter: Greening the ghetto            1      1151367060   
4  Hans Rosling: The best stats you've ever seen            1      1151440680   

   ...                                      related_talks  \
0  ...  [{'id': 865, 'hero': 'https://pe.tedcdn.com/im...   
1  ...  [{'id': 243, 'hero': 'https://pe.tedcdn.com/im...   
2  ...  [{'id': 1725, 'hero': 'https://pe.tedcdn.com/i...   
3  ...  [{'id': 1041, 'hero': 'https://pe.tedcdn.com/i...   
4  ...  [{'id': 2056, 'hero': 'https://pe.tedcdn.com/i...   

                     speaker_occupation  \
0                       Author/educator   
1                      Climate advocate   
2                  Technology columnist   
3    Activist for environmental justice   
4  Global health expert; data visionary   

                                                tags  \
0  ['children', 'creativity', 'culture', 'dance',...   
1  ['alternative energy', 'cars', 'climate change...   
2  ['computers', 'entertainment', 'interface desi...   
3  ['MacArthur grant', 'activism', 'business', 'c...   
4  ['Africa', 'Asia', 'Google', 'demo', 'economic...   

                             title  \
0      Do schools kill creativity?   
1      Averting the climate crisis   
2                 Simplicity sells   
3              Greening the ghetto   
4  The best stats you've ever seen   

                                                 url     views  \
0  https://www.ted.com/talks/ken_robinson_says_sc...  47227110   
1  https://www.ted.com/talks/al_gore_on_averting_...   3200520   
2  https://www.ted.com/talks/david_pogue_says_sim...   1636292   
3  https://www.ted.com/talks/majora_carter_s_tale...   1697550   
4  https://www.ted.com/talks/hans_rosling_shows_t...  12005869   

   comments_per_view  views_per_comment  film_datetime  \
0           0.000096       10372.745443     2006-02-25   
1           0.000083       12077.433962     2006-02-25   
2           0.000076       13195.903226     2006-02-24   
3           0.000118        8487.750000     2006-02-26   
4           0.000049       20245.984823     2006-02-22   

                                        ratings_list  
0  [{'id': 7, 'name': 'Funny', 'count': 19645}, {...  
1  [{'id': 7, 'name': 'Funny', 'count': 544}, {'i...  
2  [{'id': 7, 'name': 'Funny', 'count': 964}, {'i...  
3  [{'id': 3, 'name': 'Courageous', 'count': 760}...  
4  [{'id': 9, 'name': 'Ingenious', 'count': 3202}...  

[5 rows x 21 columns]

In [42]:
# Let's try that again
ted_event_views['views_per_speaker'] = ted_event_views.views/ted_event_views.event

In [43]:
ted_event_views.sort_values('views_per_speaker').tail()

event     views  views_per_speaker
event                                                
TEDxNorrkoping         1   6569493          6569493.0
TEDxCreativeCoast      1   8444981          8444981.0
TEDxBloomington        2  18968519          9484259.5
TEDxHouston            2  32280501         16140250.5
TEDxPuget Sound        1  34309432         34309432.0

We want the event with the highest view to talk ratio with ties being most speakers

**Some suggestions**
- count the number of `jaw_dropping` in the ratings
- count percentage of positive views based on the ratings and aggregate by event
- the number of talks should be factoredd in
- the views to positive comment ratio

**Something Simple**

In [35]:
ted.event.value_counts().head()

TED2014    84
TED2009    83
TED2016    77
TED2013    77
TED2015    75
Name: event, dtype: int64

If we want to know about talk quality we could look at views.

In [45]:
ted.groupby('event').views.mean().sort_values().tail()

event
TEDxNorrkoping        6569493.0
TEDxCreativeCoast     8444981.0
TEDxBloomington       9484259.5
TEDxHouston          16140250.5
TEDxPuget Sound      34309432.0
Name: views, dtype: float64

`Groupby`: for each unique value in a selected column (`event`), we want the aggregate function (`mean`) over some other column/s (`views`).

`TEDxPuget Sound` had a high average view count but we need to know the number of talks at this event.



In [52]:
# we are going to use sort_values which requires by=['column name'] which we can choose any of the columns we like
ted.groupby('event').views.agg(['count', 'mean']).sort_values(by=['mean']).tail()

count        mean
event                               
TEDxNorrkoping         1   6569493.0
TEDxCreativeCoast      1   8444981.0
TEDxBloomington        2   9484259.5
TEDxHouston            2  16140250.5
TEDxPuget Sound        1  34309432.0

We don't know if `TEDxPuget Sound` only had one talk but we can't evaluate it as the best event based only on one talk.

In [54]:
# we can sum the views and order by 'sum' to see which event had the most views.
ted.groupby('event').views.agg(['count', 'mean', 'sum']).sort_values(by=['sum']).tail()

count          mean        sum
event                                         
TED2006            45  3.274345e+06  147345533
TED2015            75  2.011017e+06  150826305
TEDGlobal 2013     66  2.584163e+06  170554736
TED2014            84  2.072874e+06  174121423
TED2013            77  2.302700e+06  177307937

TED2013 had the most views at 177,307,937 views...but we already noted the bias in views for older talks that have had more time to accumulate than more recent talks. Also, it is possible the the worst talks might be viewed more than more average talks. In fact the extremes of the best and the worst tend to be most interesting and therefore views don't define quality. In this case we would need to delve into the ratings column which would make our code much more complex.

**Lessons**
1. Think creatively for how tyou can use the data youhave to answer your question
    - You'll rarely have all the data you need
    - Always talk about the weakness of your approach
    - You may need to gather more data
2. Watch out for small sample sizes
    - Use count with mean to find meaningless means

# 6. Unpack the ratings data

In [55]:
ted.ratings.head()

0    [{'id': 7, 'name': 'Funny', 'count': 19645}, {...
1    [{'id': 7, 'name': 'Funny', 'count': 544}, {'i...
2    [{'id': 7, 'name': 'Funny', 'count': 964}, {'i...
3    [{'id': 3, 'name': 'Courageous', 'count': 760}...
4    [{'id': 9, 'name': 'Ingenious', 'count': 3202}...
Name: ratings, dtype: object

This comes from the TED website and ther used to be a way for viewers to tag talks with adjectives. This functionality has been depricated but the data we have refers to that evaluation system.

In [58]:
# Use the .loc to find the row you want and column you want
ted.loc[0, 'ratings']

"[{'id': 7, 'name': 'Funny', 'count': 19645}, {'id': 1, 'name': 'Beautiful', 'count': 4573}, {'id': 9, 'name': 'Ingenious', 'count': 6073}, {'id': 3, 'name': 'Courageous', 'count': 3253}, {'id': 11, 'name': 'Longwinded', 'count': 387}, {'id': 2, 'name': 'Confusing', 'count': 242}, {'id': 8, 'name': 'Informative', 'count': 7346}, {'id': 22, 'name': 'Fascinating', 'count': 10581}, {'id': 21, 'name': 'Unconvincing', 'count': 300}, {'id': 24, 'name': 'Persuasive', 'count': 10704}, {'id': 23, 'name': 'Jaw-dropping', 'count': 4439}, {'id': 25, 'name': 'OK', 'count': 1174}, {'id': 26, 'name': 'Obnoxious', 'count': 209}, {'id': 10, 'name': 'Inspiring', 'count': 24924}]"

In [59]:
# another way
ted.ratings[0]

"[{'id': 7, 'name': 'Funny', 'count': 19645}, {'id': 1, 'name': 'Beautiful', 'count': 4573}, {'id': 9, 'name': 'Ingenious', 'count': 6073}, {'id': 3, 'name': 'Courageous', 'count': 3253}, {'id': 11, 'name': 'Longwinded', 'count': 387}, {'id': 2, 'name': 'Confusing', 'count': 242}, {'id': 8, 'name': 'Informative', 'count': 7346}, {'id': 22, 'name': 'Fascinating', 'count': 10581}, {'id': 21, 'name': 'Unconvincing', 'count': 300}, {'id': 24, 'name': 'Persuasive', 'count': 10704}, {'id': 23, 'name': 'Jaw-dropping', 'count': 4439}, {'id': 25, 'name': 'OK', 'count': 1174}, {'id': 26, 'name': 'Obnoxious', 'count': 209}, {'id': 10, 'name': 'Inspiring', 'count': 24924}]"

In [60]:
# What type of object is this?
type(ted.ratings[0])

str

This is a stringified (...a string representation of a....) list of dictionaries. There are a number of things we could do but a simple and new suggestion for me is to import a module called Abstract Syntax Tree 

In [61]:
import ast

In [64]:
ast.literal_eval('[1,2,3]')

[1, 2, 3]

In [65]:
type(ast.literal_eval('[1,2,3]'))

list

We pass it a string of what looks like a list and it returns a list

In [66]:
ast.literal_eval(ted.ratings[0])

[{'id': 7, 'name': 'Funny', 'count': 19645},
 {'id': 1, 'name': 'Beautiful', 'count': 4573},
 {'id': 9, 'name': 'Ingenious', 'count': 6073},
 {'id': 3, 'name': 'Courageous', 'count': 3253},
 {'id': 11, 'name': 'Longwinded', 'count': 387},
 {'id': 2, 'name': 'Confusing', 'count': 242},
 {'id': 8, 'name': 'Informative', 'count': 7346},
 {'id': 22, 'name': 'Fascinating', 'count': 10581},
 {'id': 21, 'name': 'Unconvincing', 'count': 300},
 {'id': 24, 'name': 'Persuasive', 'count': 10704},
 {'id': 23, 'name': 'Jaw-dropping', 'count': 4439},
 {'id': 25, 'name': 'OK', 'count': 1174},
 {'id': 26, 'name': 'Obnoxious', 'count': 209},
 {'id': 10, 'name': 'Inspiring', 'count': 24924}]

In [67]:
def str_list(ratings_str):
    return ast.literal_eval(ratings_str)

In [68]:
# test our function
str_list(ted.ratings[0])

[{'id': 7, 'name': 'Funny', 'count': 19645},
 {'id': 1, 'name': 'Beautiful', 'count': 4573},
 {'id': 9, 'name': 'Ingenious', 'count': 6073},
 {'id': 3, 'name': 'Courageous', 'count': 3253},
 {'id': 11, 'name': 'Longwinded', 'count': 387},
 {'id': 2, 'name': 'Confusing', 'count': 242},
 {'id': 8, 'name': 'Informative', 'count': 7346},
 {'id': 22, 'name': 'Fascinating', 'count': 10581},
 {'id': 21, 'name': 'Unconvincing', 'count': 300},
 {'id': 24, 'name': 'Persuasive', 'count': 10704},
 {'id': 23, 'name': 'Jaw-dropping', 'count': 4439},
 {'id': 25, 'name': 'OK', 'count': 1174},
 {'id': 26, 'name': 'Obnoxious', 'count': 209},
 {'id': 10, 'name': 'Inspiring', 'count': 24924}]

In [70]:
ted.ratings.apply(str_list).head()

0    [{'id': 7, 'name': 'Funny', 'count': 19645}, {...
1    [{'id': 7, 'name': 'Funny', 'count': 544}, {'i...
2    [{'id': 7, 'name': 'Funny', 'count': 964}, {'i...
3    [{'id': 3, 'name': 'Courageous', 'count': 760}...
4    [{'id': 9, 'name': 'Ingenious', 'count': 3202}...
Name: ratings, dtype: object

In [69]:
# Actually we don't need the function because it simply does the same as the the funciton contained within it
ted.ratings.apply(ast.literal_eval).head()

0    [{'id': 7, 'name': 'Funny', 'count': 19645}, {...
1    [{'id': 7, 'name': 'Funny', 'count': 544}, {'i...
2    [{'id': 7, 'name': 'Funny', 'count': 964}, {'i...
3    [{'id': 3, 'name': 'Courageous', 'count': 760}...
4    [{'id': 9, 'name': 'Ingenious', 'count': 3202}...
Name: ratings, dtype: object

In [72]:
# Another way
ted.ratings.apply(lambda x: ast.literal_eval(x)).head()

0    [{'id': 7, 'name': 'Funny', 'count': 19645}, {...
1    [{'id': 7, 'name': 'Funny', 'count': 544}, {'i...
2    [{'id': 7, 'name': 'Funny', 'count': 964}, {'i...
3    [{'id': 3, 'name': 'Courageous', 'count': 760}...
4    [{'id': 9, 'name': 'Ingenious', 'count': 3202}...
Name: ratings, dtype: object

In [74]:
# Now save it to a new column
ted['ratings_list'] = ted.ratings.apply(ast.literal_eval)

In [76]:
# Test it
ted.ratings_list[0]

[{'id': 7, 'name': 'Funny', 'count': 19645},
 {'id': 1, 'name': 'Beautiful', 'count': 4573},
 {'id': 9, 'name': 'Ingenious', 'count': 6073},
 {'id': 3, 'name': 'Courageous', 'count': 3253},
 {'id': 11, 'name': 'Longwinded', 'count': 387},
 {'id': 2, 'name': 'Confusing', 'count': 242},
 {'id': 8, 'name': 'Informative', 'count': 7346},
 {'id': 22, 'name': 'Fascinating', 'count': 10581},
 {'id': 21, 'name': 'Unconvincing', 'count': 300},
 {'id': 24, 'name': 'Persuasive', 'count': 10704},
 {'id': 23, 'name': 'Jaw-dropping', 'count': 4439},
 {'id': 25, 'name': 'OK', 'count': 1174},
 {'id': 26, 'name': 'Obnoxious', 'count': 209},
 {'id': 10, 'name': 'Inspiring', 'count': 24924}]

In [77]:
type(ted.ratings_list[0])

list

In [78]:
ted.dtypes

comments                       int64
description                   object
duration                       int64
event                         object
film_date                      int64
languages                      int64
main_speaker                  object
name                          object
num_speaker                    int64
published_date                 int64
ratings                       object
related_talks                 object
speaker_occupation            object
tags                          object
title                         object
url                           object
views                          int64
comments_per_view            float64
views_per_comment            float64
film_datetime         datetime64[ns]
views_per_speaker            float64
ratings_list                  object
dtype: object

Confirmed that everything is as we expect....ratings_list is list object

**Lesson**
1. Pay attention to data types
    - dtypes impact everything about pandas functionality
2. Use apply whenever necessary
    - It's often discouraged because it is slow compared to built-in pandas functions/methods
    - Sometimes performace isn't inmportant
    - Use it last instead of first
    - Built-in functionsa are more reliable, better documented, faster, better tested
    - It can be confusing because there are series apply method, series map method, dataframe apply method, and dataframe applymap method

# 7. Count the total numer of ratings recieved by each talk.
Store it in a new column named `num_ratings`

Bonus:
    1. for each talk, calculate the percentage of ratings that were negative
    2. for each talk, calculate the average number of ratings received per day since it was published
   

In [122]:
type(ted.ratings_list[0])

list

We want to sum the count. This is a good time to make a function.

In [198]:
# We did some debugging as we developed this fucntion piece by piece.
def cnt_ratings(ls):
    empty = []
    for l in ls:
        cnt = 0
        for d in l:
            cnt += d.get('count')   
        empty.append(cnt) 
    return empty

In [200]:
# Want to see the length of this before inserting it into our dateframe
num_ratings = []
num_ratings = cnt_ratings(ted.ratings_list)

In [195]:
len(num_ratings)

2550

In [ ]:
ted['num_ratings'] = cnt_ratings(ted.ratings_list)

In [201]:
ted.tail()

comments                                        description  duration  \
2545        17  Between 2008 and 2016, the United States depor...       476   
2546         6  How can you study Mars without a spaceship? He...       290   
2547        10  Science fiction visions of the future show us ...       651   
2548        32  In an unmissable talk about race and politics ...      1100   
2549         8  With more than half of the world population li...       519   

             event   film_date  languages         main_speaker  \
2545       TED2017  1496707200          4     Duarte Geraldino   
2546       TED2017  1492992000          3  Armando Azua-Bustos   
2547       TED2017  1492992000          1       Radhika Nagpal   
2548  TEDxMileHigh  1499472000          1     Theo E.J. Wilson   
2549       TED2017  1492992000          1    Karoliina Korppoo   

                                                   name  num_speaker  \
2545  Duarte Geraldino: What we're missing in the de...            1   
2546  Armando Azua-Bustos: The most Martian place on...            1   
2547  Radhika Nagpal: What intelligent machines can ...            1   
2548  Theo E.J. Wilson: A black man goes undercover ...            1   
2549  Karoliina Korppoo: How a video game might help...            1   

      published_date  ...   speaker_occupation  \
2545      1505851216  ...           Journalist   
2546      1505919737  ...       Astrobiologist   
2547      1506006095  ...    Robotics engineer   
2548      1506024042  ...  Public intellectual   
2549      1506092422  ...        Game designer   

                                                   tags  \
2545  ['TED Residency', 'United States', 'community'...   
2546  ['Mars', 'South America', 'TED Fellows', 'astr...   
2547  ['AI', 'ants', 'fish', 'future', 'innovation',...   
2548  ['Internet', 'TEDx', 'United States', 'communi...   
2549  ['cities', 'design', 'future', 'infrastructure...   

                                                  title  \
2545  What we're missing in the debate about immigra...   
2546                    The most Martian place on Earth   
2547  What intelligent machines can learn from a sch...   
2548       A black man goes undercover in the alt-right   
2549  How a video game might help us build better ci...   

                                                    url   views  \
2545  https://www.ted.com/talks/duarte_geraldino_wha...  450430   
2546  https://www.ted.com/talks/armando_azua_bustos_...  417470   
2547  https://www.ted.com/talks/radhika_nagpal_what_...  375647   
2548  https://www.ted.com/talks/theo_e_j_wilson_a_bl...  419309   
2549  https://www.ted.com/talks/karoliina_korppoo_ho...  391721   

     comments_per_view  views_per_comment  film_datetime  \
2545          0.000038       26495.882353     2017-06-06   
2546          0.000014       69578.333333     2017-04-24   
2547          0.000027       37564.700000     2017-04-24   
2548          0.000076       13103.406250     2017-07-08   
2549          0.000020       48965.125000     2017-04-24   

                                           ratings_list num_ratings  
2545  [{'id': 3, 'name': 'Courageous', 'count': 24},...         192  
2546  [{'id': 22, 'name': 'Fascinating', 'count': 32...         151  
2547  [{'id': 1, 'name': 'Beautiful', 'count': 14}, ...         136  
2548  [{'id': 11, 'name': 'Longwinded', 'count': 3},...         583  
2549  [{'id': 21, 'name': 'Unconvincing', 'count': 2...         142  

[5 rows x 22 columns]

In [142]:
empty = [cnt_ratings(l) for l in ted.ratings_list]

In [146]:
empty[:5]

[19645, 544, 964, 760, 3202]

In [162]:
for l in ted.ratings_list[0:5]:
    cnt = 0
    for d in l:
        cnt += d.get('count')
    print(cnt)    

93850
2936
2824
3728
25620


In [119]:
ted = ted.drop(columns='num_ratings')

In [120]:
ted

comments                                        description  duration  \
0         4553  Sir Ken Robinson makes an entertaining and pro...      1164   
1          265  With the same humor and humanity he exuded in ...       977   
2          124  New York Times columnist David Pogue takes aim...      1286   
3          200  In an emotionally charged talk, MacArthur-winn...      1116   
4          593  You've never seen data presented like this. Wi...      1190   
...        ...                                                ...       ...   
2545        17  Between 2008 and 2016, the United States depor...       476   
2546         6  How can you study Mars without a spaceship? He...       290   
2547        10  Science fiction visions of the future show us ...       651   
2548        32  In an unmissable talk about race and politics ...      1100   
2549         8  With more than half of the world population li...       519   

             event   film_date  languages         main_speaker  \
0          TED2006  1140825600         60         Ken Robinson   
1          TED2006  1140825600         43              Al Gore   
2          TED2006  1140739200         26          David Pogue   
3          TED2006  1140912000         35        Majora Carter   
4          TED2006  1140566400         48         Hans Rosling   
...            ...         ...        ...                  ...   
2545       TED2017  1496707200          4     Duarte Geraldino   
2546       TED2017  1492992000          3  Armando Azua-Bustos   
2547       TED2017  1492992000          1       Radhika Nagpal   
2548  TEDxMileHigh  1499472000          1     Theo E.J. Wilson   
2549       TED2017  1492992000          1    Karoliina Korppoo   

                                                   name  num_speaker  \
0             Ken Robinson: Do schools kill creativity?            1   
1                  Al Gore: Averting the climate crisis            1   
2                         David Pogue: Simplicity sells            1   
3                    Majora Carter: Greening the ghetto            1   
4         Hans Rosling: The best stats you've ever seen            1   
...                                                 ...          ...   
2545  Duarte Geraldino: What we're missing in the de...            1   
2546  Armando Azua-Bustos: The most Martian place on...            1   
2547  Radhika Nagpal: What intelligent machines can ...            1   
2548  Theo E.J. Wilson: A black man goes undercover ...            1   
2549  Karoliina Korppoo: How a video game might help...            1   

      published_date  ...                                      related_talks  \
0         1151367060  ...  [{'id': 865, 'hero': 'https://pe.tedcdn.com/im...   
1         1151367060  ...  [{'id': 243, 'hero': 'https://pe.tedcdn.com/im...   
2         1151367060  ...  [{'id': 1725, 'hero': 'https://pe.tedcdn.com/i...   
3         1151367060  ...  [{'id': 1041, 'hero': 'https://pe.tedcdn.com/i...   
4         1151440680  ...  [{'id': 2056, 'hero': 'https://pe.tedcdn.com/i...   
...              ...  ...                                                ...   
2545      1505851216  ...  [{'id': 2596, 'hero': 'https://pe.tedcdn.com/i...   
2546      1505919737  ...  [{'id': 2491, 'hero': 'https://pe.tedcdn.com/i...   
2547      1506006095  ...  [{'id': 2346, 'hero': 'https://pe.tedcdn.com/i...   
2548      1506024042  ...  [{'id': 2512, 'hero': 'https://pe.tedcdn.com/i...   
2549      1506092422  ...  [{'id': 2682, 'hero': 'https://pe.tedcdn.com/i...   

                        speaker_occupation  \
0                          Author/educator   
1                         Climate advocate   
2                     Technology columnist   
3       Activist for environmental justice   
4     Global health expert; data visionary   
...                                    ...   
2545                            Journalist   
2546                        Astrobiologist   
2547                    

**In-class Solution**

In [147]:
def get_num_ratings(list_of_dicts):
    return list_of_dicts[0]

In [148]:
get_num_ratings(ted.ratings_list)

[{'id': 7, 'name': 'Funny', 'count': 19645},
 {'id': 1, 'name': 'Beautiful', 'count': 4573},
 {'id': 9, 'name': 'Ingenious', 'count': 6073},
 {'id': 3, 'name': 'Courageous', 'count': 3253},
 {'id': 11, 'name': 'Longwinded', 'count': 387},
 {'id': 2, 'name': 'Confusing', 'count': 242},
 {'id': 8, 'name': 'Informative', 'count': 7346},
 {'id': 22, 'name': 'Fascinating', 'count': 10581},
 {'id': 21, 'name': 'Unconvincing', 'count': 300},
 {'id': 24, 'name': 'Persuasive', 'count': 10704},
 {'id': 23, 'name': 'Jaw-dropping', 'count': 4439},
 {'id': 25, 'name': 'OK', 'count': 1174},
 {'id': 26, 'name': 'Obnoxious', 'count': 209},
 {'id': 10, 'name': 'Inspiring', 'count': 24924}]

In [149]:
get_num_ratings(ted.ratings_list[0])

{'id': 7, 'name': 'Funny', 'count': 19645}

In [150]:
# We build this piece by piece
def get_num_ratings(list_of_dicts):
    return list_of_dicts[0]['count']

In [152]:
get_num_ratings(ted.ratings_list[0])

19645

In [157]:
def get_num_ratings(list_of_dicts):
    num = 0
    for d in list_of_dicts:
        num += d['count']
    return num

In [158]:
get_num_ratings(ted.ratings_list[0])

93850

In [205]:
ted.ratings_list.apply(get_num_ratings)

0       93850
1        2936
2        2824
3        3728
4       25620
        ...  
2545      192
2546      151
2547      136
2548      583
2549      142
Name: ratings_list, Length: 2550, dtype: int64

We created the column eariler so we are going to check our column.

In [206]:
ted.num_ratings.describe()

count     2550.000000
mean      2436.408235
std       4226.795631
min         68.000000
25%        870.750000
50%       1452.500000
75%       2506.750000
max      93850.000000
Name: num_ratings, dtype: float64

We are seeing 2550 values, min of 68, max of 93850 (we know that is a true value from before), there aren't any negative values.

In [207]:
# This is another way to get at the same
# Much easier and cleaner
pd.DataFrame(ted.ratings_list[0])

id          name  count
0    7         Funny  19645
1    1     Beautiful   4573
2    9     Ingenious   6073
3    3    Courageous   3253
4   11    Longwinded    387
5    2     Confusing    242
6    8   Informative   7346
7   22   Fascinating  10581
8   21  Unconvincing    300
9   24    Persuasive  10704
10  23  Jaw-dropping   4439
11  25            OK   1174
12  26     Obnoxious    209
13  10     Inspiring  24924

In [208]:
pd.DataFrame(ted.ratings_list[0]).count.sum()

AttributeError: 'function' object has no attribute 'sum'

If a column name conflicts with a built in method then you cannot use dot notation.

In [210]:
pd.DataFrame(ted.ratings_list[0])['count'].sum()

93850

**Lessons**
1. Write code in small chunks and check things as you go.
2. Lambda is best for simple functions.

Before we store magic our df we are going to change the name so that we don't overwrite our stored value of ted from the previous Jupyter Notebook

In [211]:
ted_2 = ted

In [212]:
%store ted_2

Stored 'ted_2' (DataFrame)


[***Click to Continue to Part 3***](3_pandas_challenge_8.ipynb)